Purchases/Purchasing/TrialOrIntroPriceEligibilityChecker.swift

```diff
@@ -0,0 +1,143 @@
+                    }
+                    self.operationDispatcher.dispatchOnMainThread {
+                        receiveEligibility(result)
+                    }
+                }
+            }
+        }
+    }
+
+    // swiftlint:disable line_length
+    @available(iOS 12.0, macOS 10.14, tvOS 12.0, watchOS 6.2, *)
+    private func sk1ModernEligibilityHandler(maybeReceiptData data: Data,
+                                             productIdentifiers: [String],
+                                             completionBlock receiveEligibility: @escaping ReceiveIntroEligibilityBlock) {
+        introEligibilityCalculator
+            .checkTrialOrIntroductoryPriceEligibility(with: data,
+                                                      productIdentifiers: Set(productIdentifiers)) { receivedEligibility, maybeError in
+                if let error = maybeError {
+                    Logger.error(Strings.receipt.parse_receipt_locally_error(error: error))
+                    self.backend.getIntroEligibility(appUserID: self.appUserID,
+                                                     receiptData: data,
+                                                     productIdentifiers: productIdentifiers) { result, maybeAnotherError in
+                        if let intoEligibilityError = maybeAnotherError {
+                            Logger.error(Strings.purchase.unable_to_get_intro_eligibility_with_error(error: intoEligibilityError))
+                        }
+                        self.operationDispatcher.dispatchOnMainThread {
+                            receiveEligibility(result)
+                        }
+                    }
+                    return
+                }
+                var convertedEligibility: [String: IntroEligibility] = [:]
+                for (key, value) in receivedEligibility {
+                    let introEligibility = IntroEligibility(eligibilityStatus: value)
+                    convertedEligibility[key] = introEligibility
+                }
+                self.operationDispatcher.dispatchOnMainThread {
+                    receiveEligibility(convertedEligibility)
+                }
+            }
+        // swiftlint:enable line_length
+    }
+
+    @available(iOS 15.0, tvOS 15.0, watchOS 8.0, macOS 12.0, *)
+    func sk2CheckTrialOrIntroPriceEligibility(_ productIdentifiers: [String]) async -> [String: IntroEligibility] {
+        let identifiers = Set(productIdentifiers)
+        var introDict = productIdentifiers.reduce(into: [:]) { resultDict, productId in
+            resultDict[productId] = IntroEligibility(eligibilityStatus: IntroEligibilityStatus.unknown)
+        }
+        let products = await productsManager.products(withIdentifiers: identifiers)
+        for product in products {
+            guard let sk2ProductDetails = product as? SK2ProductDetails else {
+                continue
+            }
```

Comment:
if we update the return type for `productsManager.products(withIdentifiers: identifiers)` to `SK2ProductDetails`, this check won't be necessary, and more importantly, we won't be able to accidentally break it